stream_alert/rule_processor/payload.py

```diff
@@ -331,7 +332,19 @@ def pre_parse(self):
             self.raw_record['Sns']['MessageId'],
             self.raw_record['EventSubscriptionArn'])
 
-        self.pre_parsed_record = self.raw_record['Sns']['Message']
+        # check if SNS message is an Amazon S3 Notification type
+        if 'Subject' in self.raw_record['Sns'] and \
+            self.raw_record['Sns']['Subject'] == "Amazon S3 Notification":
+            LOGGER.debug('Amazon S3 notification detected.. parsing message')
+
+            message = json.loads(self.raw_record['Sns']['Message'])['Records'][0]
+            # encapsulate in s3 stream payload
+            s3_payload = load_stream_payload('s3', 'sns-passthrough', message)
+            s3_payload = s3_payload.pre_parse().next()
```

Comment:
It appears this code is assuming the file downloaded from s3 is only going to contain 1 line? 🤔 is that the expected logic here?

If not, I would suggest actually yielding all records from this, like so:


```
if self.raw_record['Sns'].get('Subject') == 'Amazon S3 Notification':
    ...
    # other logic above
    ...
    for record in s3_payload.pre_parse():
        yield record
else:
    self.pre_parsed_record = self.raw_record['Sns']['Message']
    yield self
```

Notice the `yield self` has been moved to within the `else` block as well.

Reply:
My understanding of generators is pretty limited 😅 but when I tried the method you proposed, I get the following error in a manual integration test:

```
object of type 'NoneType' has no len(): TypeError
Traceback (most recent call last):
File "/var/task/stream_alert/rule_processor/main.py", line 26, in handler
StreamAlert(context).run(event)
File "/var/task/stream_alert/rule_processor/handler.py", line 122, in run
payload_with_normalized_records.extend(self._process_alerts(payload))
File "/var/task/stream_alert/rule_processor/handler.py", line 209, in _process_alerts
len(payload.records),
TypeError: object of type 'NoneType' has no len()
```
It seems like the return payload object will have an NoneType `records` parameter